aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
   });
 }
 
+// This reduction accumulates results as the type `acc_t`. By default, when
+// `scalar_t` is complex, `acc_t` is the downgraded real number type.
+// Otherwise, `acc_t` and `scalar_t` are the same type.
+template <typename scalar_t, typename acc_t=typename scalar_value_type<scalar_t>::type, typename out_t=typename scalar_value_type<scalar_t>::type>
+void linalg_vector_norm_kernel_cuda_impl(TensorIterator& iter, Scalar val) {
+  double ord_val;
+  if (val.isIntegral(/*includeBool=*/false)) {
+     ord_val = val.to<int64_t>();
+  } else if (val.isFloatingPoint()) {
+     ord_val = val.to<double>();
+  } else {
+     TORCH_CHECK(false, "linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
+  }
+  if (iter.numel() == 0) {
+    iter.output().fill_((ord_val < 0) ? INFINITY : 0);
+    return;
+  }
+  if (ord_val == 0) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormZeroOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == 1) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormOneOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == 2) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormTwoOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == INFINITY) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMaxOps<scalar_t, acc_t>(), 0);
+  } else if (ord_val == -INFINITY) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMinOps<scalar_t, acc_t>(), std::numeric_limits<acc_t>::infinity());
+  } else {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NormOps<scalar_t, acc_t>{ static_cast<acc_t>(ord_val) }, 0);
+  }
+  // For complex outputs, the above kernels do not touch the imaginary values,
+  // so we must zero them out
+  if (isComplexType(iter.output().scalar_type())) {
```

Comment:
@ngimel points out this is a problem for torch.norm(), too

@mruberry asks if we can address this in the future to avoid the zeroing

@ngimel says we should address this in the future